### PR TITLE
add attention_mask and position_ids in assisted model

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @gante 

Do you think that we should also add `assistant_attention_mask` and `assistant_position_ids` in `assisted_decoding`? I see that the original model has `attention_mask` and `position_ids`(in most models) in the model inputs but the assistant model has no these kinds of input.

If you think it is okay to align the inputs of the original model and the assistant model, maybe we can find a more elegant way to integrate it. Thx!

### CODE REVIEW

**Comment by gante** [comment]
Hi @jiqing-feng 👋 

I agree in principle with the changes that you are proposing, but you probably need to do a few changes to make our CI go green :)

**Comment by jiqing-feng** [comment]
Hi @gante . I use `assistant_model.prepare_inputs_for_generation` to get the inputs of the assistant model. The CI all goes green and I also tested on my several examples to make sure the output is correct. Would you please help me review it? Thx!

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4447]:
```suggestion
                assistant_model_outputs = assistant_model(**assistant_inputs)
```
(side note: thank you for these changes, calling the assistant model becomes much cleaner 👍 )

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4444]: request: this change should work due to a recent PR in the models' prepare_inputs_for_generation, and the input_ids/decoder_input_ids in assistant_inputs should have .shape[-1] equal to 1 or 2. Can we raise an exception here if it is not the case?

(there may be silent failures in custom code on the hub if we don't add the exception)

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4605]:
```suggestion
            # Update attention_mask for the assistant's next round of generations
```
(add some more information)

**Review by gante — COMMENTED** [review]
Added a few nits -- after those are addressed, we're ready to merge :)

**Comment by jiqing-feng** [comment]
Hi @gante . Would you please review it again? Thx!

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
                    if input_ids_len not in (1, 2):
```

**Review by gante — APPROVED** [review]
Thank you for iterating! 💛

**Comment by gante** [comment]
@jiqing-feng  Ah, actually I have two requests before asking for the green light of a core maintainer:
1. There is a merge conflict, due to recent changes for a new model. If you're not able to sort it out, let me know :)
2. Let's confirm that we haven't lost throughput with the changes (e.g. the assertion might be producing slowdowns). To test it, feel free to clone [this folder](https://github.com/gante/huggingface-demos/tree/main/experiments/faster_generation), move there, and then run `python benchmark_decoder_open.py facebook/opt-6.7b --aux-model facebook/opt-125m --dtype fp16 --num-samples 20` on `main` and on your branch. The execution times should be nearly identical! 🤗

**Comment by jiqing-feng** [comment]
Hi @gante . I tested it on my CPU device since the GPU is unavailable to me. The new branch is a little faster than the main branch. The test script is as follows, feel free to test it on both GPU and CPU.
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import time

prompt = "Speculative decoding is"
checkpoint = "bigscience/bloom-7b1"
assistant_checkpoint = "bigscience/bloom-560m"
device = "cpu"

tokenizer = AutoTokenizer.from_pretrained(checkpoint)
inputs = tokenizer(prompt, return_tensors="pt").to(device)

model = AutoModelForCausalLM.from_pretrained(checkpoint, torch_dtype=torch.bfloat16, low_cpu_mem_usage=True).to(device)

generation_kwargs = {"do_sample": False, "max_new_tokens": 64, "temperature": 1.0, "top_p": 1.0, "num_beams": 1}

assistant_model = AutoModelForCausalLM.from_pretrained(assistant_checkpoint, torch_dtype=torch.bfloat16, low_cpu_mem_usage=True).to(device)

for i in range(5):
    start = time.time()
    outputs = model.generate(**inputs, assistant_model=assistant_model, **generation_kwargs)
    end = time.time()
    new_tokens = outputs.shape[-1] - inputs["input_ids"].shape[-1]
    print(f"Assistant decoding latency per token is {(end-start)/new_tokens * 1000} ms")
print(tokenizer.batch_decode(outputs, skip_special_tokens=True))
```

**Comment by jiqing-feng** [comment]
Hi @gante . Would you please try it again? It should be fixed and I also tested it on A100, the results and performance are exactly the same.

**Comment by gante** [comment]
@jiqing-feng perfect, all works well on my end.

Two related notes:
1. The CI is indeed red for external reasons, waiting for [this PR](https://github.com/huggingface/transformers/pull/27353) to get merged
2. The diff shows that `assistant_accepts_encoder_outputs` (a recent addition in assisted generation for distil-whisper) is removed, which means your changes are not built on top of the latest version. 
👉 you will need to rebase your changes to fix both issues, but only after the PR linked above gets merged. You may get minor rebase issues due to 2., but they should be trivial to fix

After that is done, I'll tag a core maintainer for a final quick check :)

**Comment by jiqing-feng** [comment]
> 2\. assistant_accepts_encoder_outputs

Hi @gante . I removed `assistant_accepts_encoder_outputs` because it is useless in my new changes, all inputs should be generated by `assistant_model.prepare_inputs_for_generation`.

**Comment by gante** [comment]
> Hi @gante . I removed assistant_accepts_encoder_outputs because it is useless in my new changes, all inputs should be generated by assistant_model.prepare_inputs_for_generation.

🤦 my apologies, you're absolutely right.

In that case, rebasing to get the CI green is all you need to do. Tagging a core maintainer for a quick final check :)

**Review by gante — APPROVED** [review]
Good to go, thank you for iterating with me 💛 

(Note: results also validated on my end, no slowdown nor generative performance drop)

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Beautiful PR :)  Nice to see a feature addition which results in improving the code's simplicity.

Only request before merging is a test to make sure the inputs can be correctly passed in

**Comment by jiqing-feng** [comment]
> Thanks for adding this!
> 
> Beautiful PR :) Nice to see a feature addition which results in improving the code's simplicity.
> 
> Only request before merging is a test to make sure the inputs can be correctly passed in

Hi @amyeroberts . [This script](https://github.com/huggingface/transformers/pull/26892#issuecomment-1797132296) could test the outputs before and after my changes and the outputs. I guess you mean adding a test in the tests file to run the CI, if so, would you please tell me which file should be modified to add this test? Thx

**Comment by jiqing-feng** [comment]
> Good to go, thank you for iterating with me 💛
> 
> (Note: results also validated on my end, no slowdown nor generative performance drop)

**Comment by jiqing-feng** [comment]
> @jiqing-feng perfect, all works well on my end.
> 
> Two related notes:
> 
> 1. The CI is indeed red for external reasons, waiting for [this PR](https://github.com/huggingface/transformers/pull/27353) to get merged
> 2. The diff shows that `assistant_accepts_encoder_outputs` (a recent addition in assisted generation for distil-whisper, to support assistants with shared encoders) is removed, which means your changes are not built on top of the latest version.
> 
> 👉 you will need to rebase your changes to fix both issues, but only after the PR linked above gets merged. You may get minor rebase issues due to 2., but they should be trivial to fix
> 
> After that is done, I'll tag a core maintainer for a final quick check :)

Hi @gante . I see [this PR](https://github.com/huggingface/transformers/pull/27353) you mentioned has been merged, but some of the CI are still red.

**Comment by amyeroberts** [comment]
@jiqing-feng There were some unexpected failures because of new package releases - thankfully not related to this PR! They should now be resolved on main - rebasing should fix them here. 

> Hi @amyeroberts . https://github.com/huggingface/transformers/pull/26892#issuecomment-1797132296 could test the outputs before and after my changes. I guess you mean adding a test in the tests file to run the CI, if so, would you please tell me which file should be modified to add this test? Thx

Yes, I meant to add a test to the CI runs. It looks like it should be tested in [tests/generation/test_utils.py](https://github.com/huggingface/transformers/blob/c5d7754b1140aaad95ad36c5d9ddfda5abda19f6/tests/generation/test_utils.py#L3024) - but I'll let @gante confirm

**Comment by gante** [comment]
(woops, wrong button)

**Comment by gante** [comment]
@amyeroberts not sure if we can test this feature reliably: there are no output differences, since assisted generation always outputs what the main model dictates and this PR only modifies the assistant model's inputs to be more aligned with the main model's.

What we should see on average is a higher speedup with masked inputs, but that is far guaranteed for all calls -- it would result in a very flaky test 🤔

**Comment by amyeroberts** [comment]
@gante I understand - I wasn't clear enough before. Really all I was looking for it to make sure that this can be safely used for different assistant models i.e. can I pass in a decoder-only model? How about encoder-decoder. So not speed or values, just API

**Comment by gante** [comment]
@amyeroberts we do have Mixin tests ([e.g.](https://github.com/huggingface/transformers/blob/c5037b459e117b9286c611092f38663f6cb763b0/tests/generation/test_utils.py#L1579)), so any issue regarding API should have been caught there :)

**Review by amyeroberts — APPROVED** [review]
Thanks again for adding!

**Comment by jiqing-feng** [comment]
Hi, @gante @amyeroberts . All CI are green. I think it is time to merge : )

**Comment by gante** [comment]
@jiqing-feng thank you for iterating with us and making our codebase better 💛

**Comment by VsonicV** [comment]
@amyeroberts @jiqing-feng There are currently some unexpected CI failures caused by `test_assisted_decoding_sample` (see #27351 and #27450 ). Are they related to this recently merged PR? I can see from the testing log that this PR did not run those tests involving `test_assisted_decoding_sample` during CI checking. Thanks!

**Comment by jiqing-feng** [comment]
> @amyeroberts @jiqing-feng There are currently some unexpected CI failures caused by `test_assisted_decoding_sample` (see #27351 and #27450 ). Are they related to this recently merged PR? I can see from the testing log that this PR did not run those tests involving `test_assisted_decoding_sample` during CI checking. Thanks!

Hi, @VsonicV . Sorry for the failed CI. It is weird that I can successfully run pytest in my local repo (which have updated to origin/main). I see that your CI failed at `blenderbot` and `pegasus`, but I can pass the local test by running pytest locally. Would you please update your repo and rerun the CI? Thx! If it still didn't solve your problem, just revert to my changes to see if this is my PR's error.

**Comment by VsonicV** [comment]
Hi, @jiqing-feng, thanks for the quick check. This happened exactly the same for me: I can run `pytest tests/models/blenderbot/test_modeling_blenderbot.py`, etc., successfully in my local up-to-date repo, but it failed in the CI checks. Moreoever, these CI failures not only happen at `blenderbot` and `pegasus`, it also happened for 'umt5' (in one of my previous CI tests), and `switch_transformers` and `t5` in another recent PR (see #27450 ). I asked here because this is the only recent PR that seems related to `test_assisted_decoding_sample`, but maybe the problem is somewhere else. Thanks for the help anyways!

**Comment by jiqing-feng** [comment]
> Hi, @jiqing-feng, thanks for the quick check. This happened exactly the same for me: I can run `pytest tests/models/blenderbot/test_modeling_blenderbot.py`, etc., successfully in my local up-to-date repo, but it failed in the CI checks. Moreoever, these CI failures not only happen at `blenderbot` and `pegasus`, it also happened for 'umt5' (in one of my previous CI tests), and `switch_transformers` and `t5` in another recent PR (see #27450 ). I asked here because this is the only recent PR that seems related to `test_assisted_decoding_sample`, but maybe the problem is somewhere else. Thanks for the help anyways!

I submitted a new [PR](https://github.com/huggingface/transformers/pull/27503), and all CI passed. Would you apply my PR and see if the CI is ok?

**Comment by VsonicV** [comment]
@jiqing-feng Hi, thanks for this prompt fix! I will rebase my PR and re-do the CI checks after your new PR is merged. Fingers crossed!

**Comment by patrickvonplaten** [comment]
This PR broke speculative decoding for Whisper, can we maybe revert it for now?

**Comment by patrickvonplaten** [comment]
Issue reported here: https://huggingface.co/openai/whisper-large-v3/discussions/20

**Pull request merged**
